src/service/common.ts

```diff
@@ -0,0 +1,142 @@
+
+export function createFile(output: string, filename: string): CodeGeneratorResponse.File {
+  const file = new CodeGeneratorResponse.File();
+  file.setName(filename);
+  file.setContent(output);
+  return file;
+}
+
+type CallingTypes = {
+  requestType: string
+  responseType: string
+};
+
+function getCallingTypes(method: MethodDescriptorProto, exportMap: ExportMap): CallingTypes {
+  return {
+    requestType: getFieldType(MESSAGE_TYPE, method.getInputType().slice(1), "", exportMap),
+    responseType: getFieldType(MESSAGE_TYPE, method.getOutputType().slice(1), "", exportMap),
+  };
+}
+
+function isUsed(fileDescriptor: FileDescriptorProto, pseudoNamespace: string, exportMap: ExportMap) {
+  return fileDescriptor.getServiceList().some(service => {
+    return service.getMethodList().some(method => {
+      const callingTypes = getCallingTypes(method, exportMap);
+      const namespacePackage = pseudoNamespace + ".";
+      return (
+        callingTypes.requestType.indexOf(namespacePackage) === 0 ||
+        callingTypes.responseType.indexOf(namespacePackage) === 0
+      );
+    });
+  });
+}
+
+export type ImportDescriptor = {
+  readonly namespace: string
+  readonly path: string
+};
+
+export type RPCMethodDescriptor = {
+  readonly nameAsPascalCase: string,
+  readonly nameAsCamelCase: string,
+  readonly functionName: string,
+  readonly serviceName: string,
+  readonly requestStream: boolean
+  readonly responseStream: boolean
+  readonly requestType: string
+  readonly responseType: string
+};
+
+export class RPCDescriptor {
+  private readonly grpcService: GrpcServiceDescriptor;
+  private readonly protoService: ServiceDescriptorProto;
+  private readonly exportMap: ExportMap;
+
+  constructor(grpcService: GrpcServiceDescriptor, protoService: ServiceDescriptorProto, exportMap: ExportMap) {
+    this.grpcService = grpcService;
+    this.protoService = protoService;
+    this.exportMap = exportMap;
+  }
+  get name(): string {
+    return this.protoService.getName();
+  }
+
+  get qualifiedName(): string {
+    return (this.grpcService.packageName ? `${this.grpcService.packageName}.` : "") + this.name;
+  }
+
+  get methods(): RPCMethodDescriptor[] {
+    return this.protoService.getMethodList()
+      .map(method => {
+        const callingTypes = getCallingTypes(method, this.exportMap);
+        const nameAsCamelCase = method.getName()[0].toLowerCase() + method.getName().substr(1);
+        return {
+          nameAsPascalCase: method.getName(),
+          nameAsCamelCase,
+          functionName: normaliseFieldObjectName(nameAsCamelCase),
+          serviceName: this.name,
+          requestStream: method.getClientStreaming(),
+          responseStream: method.getServerStreaming(),
+          requestType: callingTypes.requestType,
+          responseType: callingTypes.responseType,
+        };
+      });
+  }
+}
+
+export class GrpcServiceDescriptor {
```

Comment:
This file is unchanged code extracted from grpcweb.ts - except for renaming this class from `GrpcWebServiceDescriptor` to `GrpcServiceDescriptor` since it wasn't necessarily specific to grpc-web.